content/articles/2017-11-15-withr-2.1.0.Rmd

```diff
@@ -0,0 +1,234 @@
+need to remember to use an `on.exit()` call after each stateful call. In
+addition by default each `on.exit()` action will overwrite any previous
+`on.exit()` action in the same block (using `add = TRUE` prevents this). It is also
+not possible to have this cleanup performed before the block has finished.
+
+[withr](http://withr.r-lib.org) is a solution to these issues. It defines a
+[large set of
+functions](http://withr.r-lib.org/#withr---run-code-with-modified-state) for
+dealing with global settings in R, such as `with_par()`. These functions set one of
+the global settings for the duration of a block of code, then automatically
+reset it after the block is completed.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  with_par(list(col = "red", pch = 19),
+    plot(mtcars$hp, mtcars$wt)
+  )
+  par("col")
+}
+my_plot()
+par("col")
+```
+
+In addition to the `with_*` functions there are `local_*` variants which simply
+last until the end of the block they are included in. These work similar to
+`on.exit()`, but you can set the options in one call rather than two.
+
+```{r}
+par("col" = "black")
+my_plot <- function(new) {
+  local_par(list(col = "red", pch = 19))
+  plot(mtcars$hp, mtcars$wt)
+}
+my_plot()
+par("col")
+```
+
+# Graphics devices
+
+There are now a comprehensive set of functions to deal with R's builtin
+[graphics devices](http://withr.r-lib.org/reference/devices.html).
+
+These functions open a new graphics device, run some code, then automatically
+close the device. This is very useful when printing multiple plots to the same
+device.
+
+```{r}
+with_pdf("test.pdf", width = 7, height = 5, {
+  plot(mtcars$hp, mtcars$mpg)
+  plot(mtcars$wt, mtcars$mpg)
+})
+```
+
+![](test.pdf)
+
+Thanks to [Richard Cotton's](https://github.com/richierocks) great [pull
+request](https://github.com/r-lib/withr/pull/37) for this feature!
+
+## Connections
+
+There are two new functions for cleaning up connections in R.
+`with_connection()` allows you to automatically close R's file connections.
+Here we create a writable file connection, write some lines to it with
+`with_connection()`, then open a read-only connection and read the file using
+`local_connection()`.
+
+```{r}
+with_connection(list(con = file("temp", "w")), {
+  writeLines(c("foo", "bar"), con)
+})
+read_temp <- function() {
+  readLines(local_connection(file("temp", "r")))
+}
+read_temp()
+```
+```{r, echo = FALSE}
+unlink("temp")
+```
+
+`with_db_connection()` provides [DBI](http://rstats-db.github.io/DBI)
+connections to Relational Database Management Systems which automatically call
+`DBI::dbDisconnect()`. Here we create a new [SQLite](https://www.sqlite.org/)
+database, connect to it with `with_db_connection()`, and write a new table to
+it. We then create another connection with `local_db_connection()` and read
+from the table.
+
+```{r}
+db <- tempfile()
+with_db_connection(
+  list(con = DBI::dbConnect(RSQLite::SQLite(), db)), {
```

Comment:
Maybe use `:memory:` for a temporary in memory database that you don't need to clean up?

Reply:
I think the issue there was I thought the memory database would be removed when the connection ended? Or does it persist until the R session ends?

Reply:
Oh I understand - you're using the same db across the two examples